pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,321 @@
+// Copyright Authors of Cilium
+
+package k8s
+
+import (
+	"errors"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// isSelectableService returns true if the service svc can be selected by a ToServices rule
+// Normally, only services without a label selector (i.e. empty services)
+// are allowed as targets of a toServices rule.
+// This is to minimize the chances of a pod IP being selected by this rule, which might
+// cause conflicting entries in the ipcache.
+//
+// This requirement, however, is dropped for HighScale IPCache mode, because pod IPs are
+// normally excluded from the ipcache regardless. Therefore, in HighScale IPCache mode,
+// all services can be selected by ToServices.
+func (p *PolicyWatcher) isSelectableService(svc *k8s.Service) bool {
+	if svc == nil {
+		return false
+	}
+	return p.config.EnableHighScaleIPcache || svc.IsExternal()
+}
+
+// onServiceEvent processes a ServiceNotification and (if necessary)
+// recalculates all policies affected by this change.
+func (p *PolicyWatcher) onServiceEvent(event k8s.ServiceNotification) {
+	err := p.updateToServicesPolicies(event.ID, event.Service, event.OldService)
+	if err != nil {
+		p.log.WithError(err).WithFields(logrus.Fields{
+			logfields.Event:     event.Action,
+			logfields.ServiceID: event.ID,
+		}).Warning("Failed to recalculate CiliumNetworkPolicy rules after service event")
+	}
+}
+
+// updateToServicesPolicies is to be invoked when a service has changed (i.e. it was
+// added, removed, its endpoints have changed, or its labels have changed).
+// This function then checks if any of the known CNP/CCNPs are affected by this
+// change, and recomputes them by calling resolveCiliumNetworkPolicyRefs.
+func (p *PolicyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.Service) error {
+	var errs []error
+
+	// Bail out early if updated service is not selectable
+	if !(p.isSelectableService(newSVC) || p.isSelectableService(oldSVC)) {
+		return nil
+	}
+
+	// newService is true if this is the first time we observe this service
+	newService := oldSVC == nil
+	// changedService is true if the service label or selector has changed
+	changedService := !newSVC.DeepEqual(oldSVC)
+
+	// candidatePolicyKeys contains the set of policy names we need to process
+	// for this service update. By default, we consider all policies with
+	// a ToServices selector as candidates.
+	candidatePolicyKeys := p.toServicesPolicies
+	if !(newService || changedService) {
+		// If the service definition itself has not changed, and it's not the
+		// first time we process this service, we only need to check the
+		// policies which are known the select the old version of the service
```

Comment:
I think there might be a typo here:

`policies which are known the select the old version` -> `policies which are known to select the old version`?